hs-bindgen-libclang/src/HsBindgen/Clang/Internal/Results.hs

```diff
@@ -104,28 +106,35 @@ ensureNotInRange = ensureOn conv (not . simpleEnumInRange)
     conv = coerceSimpleEnum . fromIntegral
 
 {-------------------------------------------------------------------------------
-  Check wrapper result
+  Clang version error
 -------------------------------------------------------------------------------}
 
-data Unsupported = Unsupported (SimpleEnum WrapperResult) Backtrace
+-- | Clang version requirement
+data ClangVersionRequirement =
+    ClangTooOld     -- ^ Clang is too old (older than version 11)
+  | Clang11Required -- ^ We need at least Clang version 11
+  | Clang12Required -- ^ We need at least Clang version 12
+  | Clang13Required -- ^ We need at least Clang version 13
+  | Clang14Required -- ^ We need at least Clang version 14
+  | Clang15Required -- ^ We need at least Clang version 15
+  | Clang16Required -- ^ We need at least Clang version 16
+  | Clang17Required -- ^ We need at least Clang version 17
+  | Clang18Required -- ^ We need at least Clang version 18
+  deriving stock (Show, Eq, Enum, Bounded, Generic)
+
+-- | The version of Clang being used is not supported
+data ClangVersionError = ClangVersionError ClangVersionRequirement Backtrace
   deriving stock (Show)
-  deriving Exception via CollectedBacktrace Unsupported
+  deriving Exception via CollectedBacktrace ClangVersionError
 
--- | Check wrapper result
---
--- Throws 'CallFailed' on 'WrapperFailed', and 'Unsupported' if the version of
--- LLVM is unsupported.
-checkWrapperResult :: HasCallStack => IO (a, SimpleEnum WrapperResult) -> IO a
-checkWrapperResult f = do
-    (result, mSupported) <- f
-    case fromSimpleEnum mSupported of
-      Right WrapperOk ->
-        return result
-      Right WrapperFailed ->
-        callFailed mSupported
-      _otherwise -> do
-        stack <- collectBacktrace
-        throwIO $ Unsupported mSupported stack
+{-------------------------------------------------------------------------------
+  Invalid CXType error
+-------------------------------------------------------------------------------}
+
+-- | Haskell-side invalid type error
+data InvalidCXTypeError = InvalidCXTypeError Backtrace
+  deriving stock (Show)
+  deriving Exception via CollectedBacktrace InvalidCXTypeError
```

Comment:
It's a bit unfortunate that we have a new exception type now, while we use `CallFailed` everywhere else. I can see why you did it, but perhaps we can avoid it by changing `data CallFailed result` to `data CallFailed hint` instead? The only purpose for that argument really is to give a clue about what went wrong. What do you think?

Reply:
Sure, that sounds good.  Thanks!

Details:

When checking results, the `CallFailed` "hint" is the result value property that is checked, and the backtrace provides the context required to understand the exception.  I refactored to do the same in this case as well: `fromSimpleEnum (cxtKind typ)` is passed as the "hint."

`fromSimpleEnum` can return a `Left` value, so I went ahead and treat that case as an error as well.

If we need to do this anywhere else, it may be worth creating a function.  We use terminology `ensure` to check the result of an action "call," so perhaps different terminology (such as `assert`) could be appropriate.

```haskell
assertTypeValid :: HasCallStack => CXType -> IO ()
assertTypeValid typ =
    case fromSimpleEnum (cxtKind typ) of
      e@Left{}                 -> callFailed e
      e@(Right CXType_Invalid) -> callFailed e
      Right{}                  -> pure ()
```

This cannot be done (like this) in module `HsBindgen.Clang.Internal.Results` because `CXType` is defined in `HsBindgen.Clang.Core`.

Reply:
Yeah, this would make sense alongside `ensureTypeValid`. We don't do much input value checking at the moment, but indeed, something to keep an eye on.